File: crm-ui/src/app/core/services/company.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { CompanyResponse } from '../models/company.model';

const BASE_URL = `${environment.apiUrl}/companies`;

/**
 * HTTP access for the `/api/companies` resource, used by the company-
 * selection control on the contact form (requirement CON-08).
 */
@Injectable({ providedIn: 'root' })
export class CompanyService {
  private readonly http = inject(HttpClient);

  /**
   * Loads every company, ordered by name (`GET /api/companies`).
   *
   * @returns all companies
   */
  getCompanies(): Observable<CompanyResponse[]> {
    return this.http.get<CompanyResponse[]>(BASE_URL);
  }
}

File: crm-service/src/main/java/com/crm/service/CompanyService.java
package com.crm.service;

import com.crm.config.CacheNames;
import com.crm.dto.CompanyResponse;
import com.crm.entity.Company;
import com.crm.repository.CompanyRepository;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Business logic for reading {@link Company} records, used by the
 * company-selection control on the contact form (requirement CON-08).
 */
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class CompanyService {

    private final CompanyRepository companyRepository;

    /**
     * Returns all companies, ordered by name.
     *
     * Reads through the {@code companies} cache for 24 hours (see
     * {@link com.crm.config.RedisConfig}).
     *
     * @return the list of companies, or an empty list if none exist
     */
    @Cacheable(cacheNames = CacheNames.COMPANIES)
    public List<CompanyResponse> getAllCompanies() {
        // Collectors.toList() (not Stream.toList()) — see rules.md §8.5.
        return companyRepository.findAllByOrderByNameAsc().stream()
                .map(CompanyResponse::from)
                .collect(Collectors.toList());
    }
}
